File: FluentFTP.CSharpExamples/FolderMonitor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using FluentFTP;
using FluentFTP.Monitors;

namespace Examples {

	internal static class FolderMonitorExample {

		// Downloads all PDF files from a folder on an FTP server
		// when they are fully uploaded (stable)
		public static async Task DownloadStablePdfFilesAsync(CancellationToken token) {
			var conn = new AsyncFtpClient("127.0.0.1", "ftptest", "ftptest");

			await using var monitor = new BlockingAsyncFtpMonitor(conn, new List<string> { "path/to/folder" });

			monitor.PollInterval = TimeSpan.FromMinutes(5);
			monitor.WaitForUpload = true;

			monitor.ChangeDetected = (static async (e) => {
				foreach (var file in e.Added
				                      .Where(x => Path.GetExtension(x) == ".pdf")) {
					var localFilePath = Path.Combine(@"C:\LocalFolder", Path.GetFileName(file));
					await e.AsyncFtpClient.DownloadFile(localFilePath, file, token: e.CancellationToken);
					await e.AsyncFtpClient.DeleteFile(file); // don't cancel this operation
				}
			});

			await conn.Connect(token);
			await monitor.Start(token);
		}

		// How to use the monitor in a console application
		public static async Task MainAsync() {
			using var tokenSource = new CancellationTokenSource();
			Console.CancelKeyPress += (_, e) =>
			{
				e.Cancel = true; // keep running until monitor is stopped
				tokenSource.Cancel(); // stop the monitor
			};

			await DownloadStablePdfFilesAsync(tokenSource.Token);
		}
	}
}

File: FluentFTP.VBExamples/FolderMonitor.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks
Imports FluentFTP
Imports FluentFTP.Monitors

Namespace Examples
	Friend Module FolderMonitorExample
		Async Function DownloadStablePdfFilesAsync(ByVal token As CancellationToken) As Task
			Dim conn = New AsyncFtpClient("127.0.0.1", "ftptest", "ftptest")
			Await conn.Connect(token)

			Using monitor = New BlockingAsyncFtpMonitor(conn, New List(Of String) From {
				"path/to/folder"
			})
				monitor.PollInterval = TimeSpan.FromMinutes(5)
				monitor.WaitForUpload = True

				monitor.ChangeDetected =
					Async Function(e)
						If True Then
							For Each file In e.Added.Where(Function(x) Path.GetExtension(x) = ".pdf")
								Dim localFilePath = Path.Combine("C:\LocalFolder", Path.GetFileName(file))
								Await e.AsyncFtpClient.DownloadFile(localFilePath, file, token:=e.CancellationToken)
								Await e.AsyncFtpClient.DeleteFile(file)
							Next
						End If
					End Function

				Await monitor.Start(token)
			End Using
		End Function

	End Module
End Namespace

